syntax = "proto3";

import "google/protobuf/timestamp.proto";
import "google/protobuf/empty.proto";

package api;
option go_package = "./api";

// UserService messages
message RegisterRequest {
  string requestId = 1;
  string username = 2;
  string password = 3;
}

message RegisterResponse {
  string jwt = 1;
}

message LoginRequest {
  string requestId = 1;
  string username = 2;
  string password = 3;
}

message LoginResponse {
  string jwt = 1;
}

// ExpenseReader messages
message GetReportRequest {
  string jwt = 1;
}

message GetReportResponse {
  map<string, double> report = 1;
}

message GetBillsRequest {
  string jwt = 1;
  string category = 2;
}

message Bill {
  double amount = 1;
  string name = 2;
  google.protobuf.Timestamp ts = 3;
}

message GetBillsResponse {
  repeated Bill bills = 1;
}

// ExpensePublisher messages
message BillMessage {
  string name = 1;
  double amount = 2;
  string category = 3;
  google.protobuf.Timestamp timestamp = 5;
  string jwt = 6;
}

// Combined service
service Api {
  // UserService methods
  rpc Register (RegisterRequest) returns (RegisterResponse) {}
  rpc Login (LoginRequest) returns (LoginResponse) {}

  // ExpenseReader methods
  rpc GetReport (GetReportRequest) returns (GetReportResponse) {}
  rpc GetBills (GetBillsRequest) returns (GetBillsResponse) {}

  // ExpensePublisher methods
  rpc CreateBill (BillMessage) returns (google.protobuf.Empty) {}
}